### ISSUE TITLE
Does not load on iPad

### ISSUE THREAD

**Issue by garrettgu10** [tetrio/issues#425]
### Describe the bug/issue
The site does not load on an iPad.

### Reproducing
Open Chrome or Safari on an iPad. Navigate to tetr.io. The page is stuck on a black screen.

### Expected behavior
I expected the website to work on iPad.

### Your setup
 - **OS and version**: iPadOS 13
 - **Browser and version**: iPad Chrome

### Additional context
May be related to #134, since all iPad browsers are required to use the webkit engine, which borrows heavily from MacOS safari.

**Comment by Kiwi357** [comment]
I’ll be willing to run any tests or troubleshooting. I want to be invested in this app for ipad/macOS Safari

**Comment by Tenchi2xh** [comment]
Safari is the Internet Explorer of modern times, it's extremely outdated and quirky and would take a looot of effort to make Tetrio work there, and even more to keeping it working after each update, I don't think it's possible for a one person dev team

Moreover, Chrome on iOS devices is just a wrapper around Safari because Apple does not allow for third-party web engines

**Comment by Kiwi357** [comment]
AH, well there goes my hopes and dreams of playing tetrio on the go. Would the possibility of an iPad app ever work either. I’d assume no because of the “one man dev team” thing. 

I’d even pay some Patreon. :D

**Comment by isametry** [comment]
>Would the possibility of an iPad app ever work either. I’d assume no because of the “one man dev team” thing.

The easier way to bring tetr.io to iPad really is supporting it on Safari. iPad Safari is actually Mac Safari under the hood.

**Comment by isametry** [comment]
> it would take a looot of effort to make Tetrio work there, and even more to keeping it working after each update, I don't think it's possible for a one person dev team

I’d like to hear this from the actual dev.

And tetr.io is literally the **only** website i’ve ever seen that doesn’t work on Safari, i’m wondering what’s so special about it? maybe i’m just incredibly, incredibly lucky.

**Comment by garrettgu10** [comment]
It seems like some codec is not supported in safari, based on console output.

**Comment by LayneTheEpic** [comment]
Plus, there's a high chance that touch support isn't implemented

**Comment by benz-phan** [comment]
lmao @garrettgu10  how can you think that Tetr.io going to load on an ipad.

**Comment by isametry** [comment]
> Plus, there's a high chance that touch support isn't implemented

No need, iPad supports keyboard & mouse. i regularly play Jstris on my iPad.

**Comment by hf02** [comment]
sorry if this isn't good etiquette on github, im new.

safari doesn't support ogg files, and all of the tetr.io sound effects are stored in a .ogg file. because of this, no loaded events for the ogg file fire because safari doesn't support ogg files, so the game doesn't start and hangs at the black screen. enabling the "disable sound entirely" option in settings will still try to load the ogg file, which still causes it to hang. you can recreate this issue on chrome desktop by going into devtools and request blocking `tetr.io/sfx/tetrio.ogg`.

tl;dr: safari doesn't support ogg audio files, tetr.io is dependent on a ogg audio file, so the game doesn't load.

**Comment by isametry** [comment]
Hello @hf02, thanks for the insight! So if what you’re saying is correct, seems to me that the solution is really easy: the game simply checks whether it’s being launched in Safari, and if that’s the case, all attempts to load .ogg files are skipped and sounds are automatically disabled.

I’d much rather live without sound than without tetr.io as a whole.

By the way, why does the launch of the game depend on the successful load of a ***sound file?*** I am no expert but that sounds like a sign of pretty bad code?

**Comment by Dimantio** [comment]
I wouldn't automatically assume the code is bad based on a codec issue @isametry A one man dev team can't possibly support every browsers/os on the first try.

**Comment by szymonszl** [comment]
A sound file is an **asset** of a game - it's no different from the mino textures or even the code itself. Depending on all assets for a proper load is *never* a bad thing, most games will error or fail to load without them.

Ignoring the lack of an asset would be quite bad, since people would have a worse experience, possibly without even knowing. Sure, you can mute the sound yourself if you don't want it - but you can do the same to a movie, and would you not be pissed off if you went to a cinema and it turned out to have no sound?

If you'd want to be picky about the code, then the error could be more visible, instead of just quietly appearing in the console.

**Comment by isametry** [comment]
**Response to** @szymonszl:

> it's no different from the mino textures or even the code itself

Except that it is? Texture files or the game engine are obviously crucial to the gameplay, but sound effects are not. The game can be played just fine without it.

Tying the launch of the game to the successful load of texture files makes absolute sense, because textures are **crucial** to the game even functioning.  

How can you compare tetris sounds effects to a movie’s audio?? That’s pretty ridiculous.

> people would have a worse experience, possibly without even knowing

There’s nothing easier than showing a big red alert window that says “warning: sound effects aren’t supported in this browser”.

**Response to** @Dimantio:

Yes i understand that. And I would’ve understood any other reason why they game doesn’t work in Safari. But tying the launch of the game *to an .ogg file* is something I just can’t wrap my head around.

**Comment by szymonszl** [comment]
> Tying the launch of the game to the successful load of texture files makes absolute sense, because textures are crucial to the game even functioning. Sounds are definitely not.

Sounds are just as much a part of the game as the textures are. They are both parts of a game's design and, more importantly, part of the experience. If you reduce the game to its mechanics, then you'd have to admit that Wolfenstein 3D, Half Life, and Halo are one game.

> How can you compare tetris sounds effects to a movie’s audio?? That’s pretty ridiculous.

Because both are technically optional. You can just watch the movie with subtitles and you can just play tetr.io muted, but in both cases you don't get the full experience.

> There’s nothing easier than showing a big red alert window that says “warning: sound effects aren’t supported in this browser”.

i'm going to pass on the "nothing easier than"

> But tying the launch of the game to an .ogg file is something I just can’t wrap my head around.

As a quick experiment, I removed the sound file from another game I had laying around.
It crashed with the following:
```
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.FileNotFoundException: Could not find file "/[...]/Stardew Valley/game/Content/XACT/Sound Bank.xsb".
File name: '/[..]/Stardew Valley/game/Content/XACT/Sound Bank.xsb'
  at System.IO.FileStream..ctor (System.String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean anonymous, FileOptions options) <0x40364000 + 0x0065b> in <filename unknown>:0 
  at System.IO.FileStream..ctor (System.String path, FileMode mode, FileAccess access, FileShare share) <0x40353a00 + 0x0004b> in <filename unknown>:0 

  ....
```
As I said, nothing weird about crashing on lacking assets.

**Comment by hf02** [comment]
@isametry 
> the game simply checks whether it’s being launched in Safari, and if that’s the case, all attempts to load .ogg files are skipped and sounds are automatically disabled.

instead of that, you could convert the .ogg file into a .mp3 file. then it would work on safari.

**Comment by isametry** [comment]
@szymonszl comparing apples to oranges really seems to be your thing. 

You’re using Stardew Valley as an example, which is an:
1. ambient,
2. content-rich,
3. natively-running game.

Now compare it to Tetris, a game that is:
1. abstract
2. relatively simple
3. running in a browser.

But for what it’s worth, it still proves your point: **developers DO launch their games dependent on audio assets.** All I’m saying is that for an online abstract puzzle game, that doesn’t make any sense.

**Comment by UniQMG522** [comment]
It's highly unlikely that failing to load sound effects is the only issue tetrio has on safari. Safari is lagging behind on modern web API support in many areas.

> There’s nothing easier than showing a big red alert window that says “warning: sound effects aren’t supported in this browser”.

The easiest solution is to just fail to load, since that doesn't require extra error case handling. It might not be pretty, but osk has a lot to deal with at the moment and it's probably not a priority to deal with individual assets failing to load.

> instead of that, you could convert the .ogg file into a .mp3 file. then it would work on safari.

`mp3`s have some oddities which make them unsuitable for sound effects - they have an unremovable gap at the beginning and end as a side effect of the way they're encoded. You could use other formats, but those have their own problems.

**Comment by o5k** [comment]
![](https://kagari.moe/:/mamiko_rika.png)

**Comment by o5k** [comment]
By the way, this is for the most part same as #134 . Ill keep both up, the main todo here is to serve another audio file as well as ogg, and then see what other fixes need to be done. I do not personally have any Apple devices so I can;t really test this though.........

**Comment by garrettgu10** [comment]
I'm willing to help with testing! I have Apple devices and webdev experience.

**Comment by nurumaik** [comment]
oh cmon just use some js library that can play ogg in safari. Like this one https://github.com/brion/ogv.js/

**Comment by isametry** [comment]
@nurumaik Right, either that, or like o5k said, convert the files into _aiff_ or _wav_ or whatever and serve that alongside the _ogg_ files.

Whichever way, it doesn’t really seem like an issue we should spend 4 months talking about.

**Comment by isametry** [comment]
It is still relevant. @o5k could you please give this a **”never-stale”**?

Anyway, since you’ve already been pinged… do you have any updates on the progress of this? :)

**Comment by ToxicToucan** [comment]
I would also love to know the progress. No rush, right now I use Moonlight to stream the game from my PC.

**Comment by isametry** [comment]
Since this issue is apparently unsolvable, may I suggest at least a "load without sounds" option to be added? 

I don't care about having sound effects in my Tetris game, what I do care about however is Chromium hogging up my resources.

For example an a redirect URL, like http://tetr.io/nosound, would be a subtle but intuitive solution.

**Comment by Chevanco29** [comment]
Im sorry if I sound rude but there are 2 options to fix tetr.io for safari: 
1. If tetr.io recognises that it is running on safari it will run a different version of the game that uses mp3 files instead of ogg files (which is harder to do).  

or 

2. Do as @isametry said and make a option for safari users to just get rid of the ogg files and have no sound.

**Comment by o5k** [comment]
i dont think you quite understand
safari is a horrid browser to target. safari not supporting a _very_ standard audio format is only the tip of the iceberg of huge fixes that would have to be done, which i cannot do as i do not have any iDevices, as well as a general downgrade in visual fidelity because safari doesn't support anything
furthermore, for now I am not interested in targeting a mobile demographic. tetr.io is already enough of an "io game", don't need to literally make it a mobile game. i mainly care about this issue for those players who use safari desktop as their main browser - but this issue has been on hold for a long time because i do not have the time or resources to expand to that very small group of people, and the amount of headaches it would bring are worrying
and that's not to speak of the other headaches of having to make the UI touchfriendly, at which point it'd also need to be portraitmode friendly, etc... it's a headache, for a demographic i am not targeting

**Comment by isametry** [comment]
I’m sorry @o5k but nobody, at any point during this discussion, mentioned phones or touch controls? I think all will agree that such a request would be way out of scope.

I wanna repeat what I said earlier—even if this ticket is about iPad, the real problem here is desktop Safari. Because iPads run desktop Safari now, so to fix it for Mac would also be fixing it for iPad.

The way I see it, I think you can lock or close down #425, because at the root of it is #134 and I personally don’t see much sense in keeping both up.

**Comment by iloverenge** [comment]
i dont think you quite understand
safari is a horrid browser to target. safari not supporting a very standard audio format is only the tip of the iceberg of huge fixes that would have to be done, which i cannot do as i do not have any iDevices, as well as a general downgrade in visual fidelity because safari doesn't support anything
i mainly care about this issue for those players who use safari desktop as their main browser - but this issue has been on hold for a long time because i do not have the time or resources to expand to that very small group of people, and the amount of headaches it would bring are worrying

**Comment by isametry** [comment]
@iloverenge Yeah thank you I’m not blind. That’s why I said that the discussion is more relevant to #134 than here.

**Comment by iloverenge** [comment]
I'm glad to hear you are not blind! 👍

**Comment by Chevanco29** [comment]
(sorry if I am not polite) I'm no expert by any means but I suggest to look at this: https://stackoverflow.com/questions/38581887/safari-doesnt-play-ogg-files-so-how-does-it-work-on-wiktionary-org

**Comment by hf02** [comment]
https://user-images.githubusercontent.com/55464333/160914677-b27e4fcb-bdd1-470d-9a8b-ba2fbc04731b.mp4

(plz ignore bad gameplay thx)

I came across something called [Playwright](https://playwright.dev/), which has the ability to run WebKit on Windows.

Trying it out, TETR.IO also fails to load from the same issue. So, I've created a script that intercepts `Audio`. If TETR.IO tries to fetch a `.ogg` file, it will instead point to the main menu music and mute it. The only issues I've seen, besides the lack of sound effects, seem to be specific to Playwright.

Trying this on my iPhone, TETR.IO gets stuck on loading the SFX again. I'm not sure as to why, and I can't check using developer tools as I don't have a Mac.

[The script with instructions are here.](https://gist.github.com/hf02/312650309a28c2434f9834633803125f)

**Comment by hf02** [comment]
**Volume warning:**

https://user-images.githubusercontent.com/55464333/160925430-ea5dd2dd-520e-4f24-8f1d-b03198cd036a.mp4

I got it to work on my iPhone, and minus the sound effects (You can always just turn the SFX slider down to 0%), TETR.IO seems to work with no issues. I assume this would also work with the Safari on iPads and Macs as well.

**Comment by o5k** [comment]
Next patch will include a sound format for special snowflake Safari. Can't guarantee everything will work well, as I cannot debug Safari, but it should allow the game to load.

**Comment by Kiwi357** [comment]
I can confirm that it works on Safari (including the iPad Pro). Thanks for the support.
